Сделал 3 ДЗ, поиск простых чисел реализовал без алгоритмов с запомина…

ikr0ni opened the pull request:
ДЗ 3 готова

Commit:
Сделал 3 ДЗ, поиск простых чисел реализовал без алгоритмов с запоминанием

v-lozhkin commented on lesson3/main.go:
```diff
@@ -0,0 +1,121 @@
+			if i%j == 0 {
+				break
+			}
+			if j == i-1 {
```
Не совсем понимаю, зачем это условие. 
Оптимизация на одно сравнение? Если да, то по моему мнению, оно того не стоит

ikr0ni replied:
i%j == 0 {break} я использовал как показатель, если делится без остатка, то не простое, 
j == i-1 это проверка добавлена специально, так как число само на себя делится без остатка, но это не делает его не простым

v-lozhkin replied:
а почему просто не сделать цикл по j до i-1?

v-lozhkin replied:
точнее оно так уже и есть. Условие j<i предусматривает, что мы не будем проверять деление i на i

ikr0ni replied:
Тогда нужно использовать переменную дополнительную, в которой хранить предположительную информацию, что число простое, которую нужно будет проверять после конца каждого внутреннего цикла, тут же от выделения памяти на эту переменную я отказался, т.е. переменная сразу добавляется в слайс по данному условию, без дополнительных переменных.

v-lozhkin replied:
Получается экономия на спичках. Экономия в одном буле, а читается уже сложнее. 
В целом ок, так тоже принято.

v-lozhkin commented on lesson3/main.go:
```diff
@@ -0,0 +1,121 @@
+	case "/":
+		if second == 0 {
+			return 0, errors.New("Devide by zero, in hell you will be in a good company")
+		} else {
```
Можно без  else тут

ikr0ni replied:
поправил тут https://github.com/ikr0ni/go_level1/pull/3/commits/e1d24d385da060bd7c6e1b1f944fe26890e24594

v-lozhkin commented on lesson3/main.go:
```diff
@@ -0,0 +1,121 @@
+			log.Fatal(err)
+		}
+	case "getPrime":
+		if *first < 3 {
```
Почему? Кажется, при вводе 0-2 вполне можно возвращать пустой массив

ikr0ni replied:
2 - простое число, но я просто не вижу смысла выводить такой массив, да и за одно еще один вывод в log)))

v-lozhkin replied:
Ну, вообще говоря, программа ведь решает определенную задачу. Не нужно додумывать за пользователя. Может, не каждый знает что 2 - это просто число?)

ikr0ni replied:
В https://github.com/ikr0ni/go_level1/pull/3/commits/e1d24d385da060bd7c6e1b1f944fe26890e24594 добавил информирование спрашивающего об этом факте))

v-lozhkin commented on lesson3/main.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+//Калькулятор на операции +,-,*,/,sqrt,^
+func simpleCalc(first int, second int, oper string) (float64, error) {
```
Если мы все равно во всех случаях приводим int к float64, то почему бы изначально не принимать их на вход функции?

ikr0ni replied:
На ввод мы принимаем только int, в принципе нет никакой разнице же где переводить int во float, до вызова функции или после, ну т.е. по факту, если здесь в функции поставить принимать на вход float, то приводить к float придется при вызове функции

v-lozhkin replied:
С точки зрения исполнения - да. 
Однако с точки зрения чтения кода, сейчас приведение происходит в N местах (каждый из операторов), а если приводить перед входом в функцию, то приведение будет только в двух местах (по одному на каждый аргумент)

ikr0ni replied:
резонно, поправил в https://github.com/ikr0ni/go_level1/pull/3/commits/e1d24d385da060bd7c6e1b1f944fe26890e24594

Commit:
1. В ошибку по поводу получения простых чисел при введении числа меньше 3 добавлен вывод, что простое число только 2
2. Перевод int во float перенесено в вызов функции simpleCalc
3. Переводы int во float убраны из функции simpleCalc
4. Убран else в case деления

v-lozhkin reviewed (approved):
Ок
